In [8]:
library(caret)

Loading required package: lattice

Loading required package: ggplot2



In [2]:
df = read.csv('diabetes.csv')

In [4]:
head(df)

Glucose BloodPressure SkinThickness Insulin BMI  DiabetesPedigreeFunction Age
1 148     72            35              0     33.6 0.63                     50 
2  85     66            29              0     26.6 0.35                     31 
3 183     64             0              0     23.3 0.67                     32 
4  89     66            23             94     28.1 0.17                     21 
5 137     40            35            168     43.1 2.29                     33 
6 116     74             0              0     25.6 0.20                     30 
  Outcome
1 1      
2 0      
3 1      
4 0      
5 1      
6 0

In [3]:
summary(df)

    Glucose      BloodPressure    SkinThickness      Insulin     
 Min.   :  0.0   Min.   :  0.00   Min.   : 0.00   Min.   :  0.0  
 1st Qu.: 99.0   1st Qu.: 62.00   1st Qu.: 0.00   1st Qu.:  0.0  
 Median :117.0   Median : 72.00   Median :23.00   Median : 30.5  
 Mean   :120.9   Mean   : 69.11   Mean   :20.54   Mean   : 79.8  
 3rd Qu.:140.2   3rd Qu.: 80.00   3rd Qu.:32.00   3rd Qu.:127.2  
 Max.   :199.0   Max.   :122.00   Max.   :99.00   Max.   :846.0  
      BMI        DiabetesPedigreeFunction      Age           Outcome     
 Min.   : 0.00   Min.   :0.0800           Min.   :21.00   Min.   :0.000  
 1st Qu.:27.30   1st Qu.:0.2400           1st Qu.:24.00   1st Qu.:0.000  
 Median :32.00   Median :0.3700           Median :29.00   Median :0.000  
 Mean   :31.99   Mean   :0.4724           Mean   :33.24   Mean   :0.349  
 3rd Qu.:36.60   3rd Qu.:0.6300           3rd Qu.:41.00   3rd Qu.:1.000  
 Max.   :67.10   Max.   :2.4200           Max.   :81.00   Max.   :1.000  

In [5]:
df$Outcome = as.factor(df$Outcome)

In [6]:
summary(df)

    Glucose      BloodPressure    SkinThickness      Insulin     
 Min.   :  0.0   Min.   :  0.00   Min.   : 0.00   Min.   :  0.0  
 1st Qu.: 99.0   1st Qu.: 62.00   1st Qu.: 0.00   1st Qu.:  0.0  
 Median :117.0   Median : 72.00   Median :23.00   Median : 30.5  
 Mean   :120.9   Mean   : 69.11   Mean   :20.54   Mean   : 79.8  
 3rd Qu.:140.2   3rd Qu.: 80.00   3rd Qu.:32.00   3rd Qu.:127.2  
 Max.   :199.0   Max.   :122.00   Max.   :99.00   Max.   :846.0  
      BMI        DiabetesPedigreeFunction      Age        Outcome
 Min.   : 0.00   Min.   :0.0800           Min.   :21.00   0:500  
 1st Qu.:27.30   1st Qu.:0.2400           1st Qu.:24.00   1:268  
 Median :32.00   Median :0.3700           Median :29.00          
 Mean   :31.99   Mean   :0.4724           Mean   :33.24          
 3rd Qu.:36.60   3rd Qu.:0.6300           3rd Qu.:41.00          
 Max.   :67.10   Max.   :2.4200           Max.   :81.00          

In [10]:
model = preProcess(df,method = 'range')

In [11]:
train_df = predict(model,newdata = df)

In [15]:
summary(train_df)

    Glucose       BloodPressure    SkinThickness       Insulin       
 Min.   :0.0000   Min.   :0.0000   Min.   :0.0000   Min.   :0.00000  
 1st Qu.:0.4975   1st Qu.:0.5082   1st Qu.:0.0000   1st Qu.:0.00000  
 Median :0.5879   Median :0.5902   Median :0.2323   Median :0.03605  
 Mean   :0.6075   Mean   :0.5664   Mean   :0.2074   Mean   :0.09433  
 3rd Qu.:0.7048   3rd Qu.:0.6557   3rd Qu.:0.3232   3rd Qu.:0.15041  
 Max.   :1.0000   Max.   :1.0000   Max.   :1.0000   Max.   :1.00000  
      BMI         DiabetesPedigreeFunction      Age           Outcome   
 Min.   :0.0000   Min.   :0.00000          Min.   :0.0000   Class0:500  
 1st Qu.:0.4069   1st Qu.:0.06838          1st Qu.:0.0500   Class1:268  
 Median :0.4769   Median :0.12393          Median :0.1333               
 Mean   :0.4768   Mean   :0.16769          Mean   :0.2040               
 3rd Qu.:0.5455   3rd Qu.:0.23504          3rd Qu.:0.3333               
 Max.   :1.0000   Max.   :1.00000          Max.   :1.0000               

In [13]:
head(train_df)

Glucose   BloodPressure SkinThickness Insulin   BMI      
1 0.7437186 0.5901639     0.3535354     0.0000000 0.5007452
2 0.4271357 0.5409836     0.2929293     0.0000000 0.3964232
3 0.9195980 0.5245902     0.0000000     0.0000000 0.3472429
4 0.4472362 0.5409836     0.2323232     0.1111111 0.4187779
5 0.6884422 0.3278689     0.3535354     0.1985816 0.6423249
6 0.5829146 0.6065574     0.0000000     0.0000000 0.3815201
  DiabetesPedigreeFunction Age       Outcome
1 0.23504274               0.4833333 Class1 
2 0.11538462               0.1666667 Class0 
3 0.25213675               0.1833333 Class1 
4 0.03846154               0.0000000 Class0 
5 0.94444444               0.2000000 Class1 
6 0.05128205               0.1500000 Class0

In [12]:
levels(train_df$Outcome) = c('Class0','Class1')

In [19]:
mod1 = glm(Outcome ~ .-Outcome, data = train_df, family = 'binomial')

In [21]:
summary(mod1)


Call:
glm(formula = Outcome ~ . - Outcome, family = "binomial", data = train_df)

Deviance Residuals: 
    Min       1Q   Median       3Q      Max  
-2.5802  -0.7285  -0.4332   0.7438   2.9185  

Coefficients:
                         Estimate Std. Error z value Pr(>|z|)    
(Intercept)              -7.65947    0.67748 -11.306  < 2e-16 ***
Glucose                   6.88437    0.72945   9.438  < 2e-16 ***
BloodPressure            -1.49034    0.63819  -2.335  0.01953 *  
SkinThickness             0.08276    0.68338   0.121  0.90361    
Insulin                  -1.15890    0.77140  -1.502  0.13301    
BMI                       5.89540    0.99659   5.916 3.31e-09 ***
DiabetesPedigreeFunction  2.08405    0.69220   3.011  0.00261 ** 
Age                       1.95961    0.48657   4.027 5.64e-05 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 993.48  on 767  degrees of freedom
Residual devian

- Diabetes depends on the glucose levels of the person
- Diabetes depends on the BloodPressure levels of the person
- Diabetes **does not** depend on SkinThickness of the person
- Diabetes <strong>does not</strong> depend on the Insulin level of the person
- Diabetes depends on the BMI of the person
- Diabetes depends on the DiabetesPedigreeFunction of the person
- Diabetes depends on the Age of the person

In [29]:
head(predict(mod1,train_df,type = 'response'))

1          2          3          4          5          6 
0.73093196 0.06934637 0.69453973 0.04986430 0.92672670 0.13006140

In [68]:
ans = as.data.frame(predict(mod1,train_df,type = 'response'))

In [69]:
ans = round(ans*100,2)

In [70]:
head(ans)

predict(mod1, train_df, type = "response")
1 73.09                                     
2  6.93                                     
3 69.45                                     
4  4.99                                     
5 92.67                                     
6 13.01

In [71]:
colnames(ans) = 'Probability'

In [82]:
ans_str = strsplit(toString(unlist(ans$Probability)),',')

In [83]:
ans_str = as.data.frame(sapply(ans_str, paste,'%',sep=''))

In [87]:
colnames(ans_str) = 'Probability'

In [88]:
head(ans_str)

Probability
1 73.09%     
2  6.93%     
3  69.45%    
4  4.99%     
5  92.67%    
6  13.01%

# Loan data

In [94]:
library(readxl)

In [95]:
df = read_excel('BankLoan.xlsx')

In [96]:
head(df)

Married Dependents Education    Self_Employed ApplicantIncome
1 Yes     1          Graduate     No            4583           
2 Yes     0          Graduate     Yes           3000           
3 Yes     0          Not Graduate No            2583           
4 No      0          Graduate     No            6000           
5 Yes     2          Graduate     Yes           5417           
6 Yes     0          Not Graduate No            2333           
  CoapplicantIncome LoanAmount Loan_Amount_Term Credit_History Property_Area
1 1508              128        360              1              Rural        
2    0               66        360              1              Urban        
3 2358              120        360              1              Urban        
4    0              141        360              1              Urban        
5 4196              267        360              1              Urban        
6 1516               95        360              1              Urban        
  Loan_Status
1 N          
2 Y          
3 Y          
4 Y          
5 Y          
6 Y

In [158]:
df_1 = df[,c(10,11)]

In [163]:
head(df_1)

Property_Area Loan_Status
1 Rural         N          
2 Urban         Y          
3 Urban         Y          
4 Urban         Y          
5 Urban         Y          
6 Urban         Y

In [165]:
library(data.table)


Attaching package: 'data.table'


The following objects are masked from 'package:dplyr':

    between, first, last




In [167]:
df_1[,":="(Loan_Status=ifelse(Loan_Status=="N",0,1))]

ERROR: Error in `:=`(Loan_Status = ifelse(Loan_Status == "N", 0, 1)): Check that is.data.table(DT) == TRUE. Otherwise, := and `:=`(...) are defined for use in j, once only and in particular ways. See help(":=").


In [168]:
library(dplyr)

In [174]:
df_1$Loan_Status = ifelse(df_1$Loan_Status == "N",0,1)

In [211]:
head(df_1,10)

Property_Area Loan_Status
1  Rural         0          
2  Urban         1          
3  Urban         1          
4  Urban         1          
5  Urban         1          
6  Urban         1          
7  Semiurban     0          
8  Urban         1          
9  Semiurban     0          
10 Urban         1

In [150]:
df_1$Property_Area = as.factor(df_1$Property_Area)

In [151]:
df_1$Loan_Status = as.numeric(as.factor(df_1$Loan_Status))

In [152]:
head(df_1)

Property_Area Loan_Status
1 Rural         1          
2 Urban         2          
3 Urban         2          
4 Urban         2          
5 Urban         2          
6 Urban         2

In [157]:
unique(df_1$Loan_Status)

[1] 1 2

In [154]:
levels(df_1$Property_Area)

[1] "Rural"     "Semiurban" "Urban"

In [153]:
contrasts(df_1$Property_Area)

Semiurban Urban
Rural     0         0    
Semiurban 1         0    
Urban     0         1

In [139]:
levels(df_1$Property_Area)

[1] "Rural"     "Semiurban" "Urban"

In [113]:
df_ = model.matrix(~Property_Area, data = df_1)

In [114]:
head(df_)

(Intercept) Property_AreaSemiurban Property_AreaUrban
1 1           0                      0                 
2 1           0                      1                 
3 1           0                      1                 
4 1           0                      1                 
5 1           0                      1                 
6 1           0                      1

In [117]:
df_1 = as.data.frame(df_1)

In [119]:
str(df_1)

'data.frame':	492 obs. of  2 variables:
 $ Property_Area: Factor w/ 3 levels "Rural","Semiurban",..: 1 3 3 3 3 3 2 3 2 3 ...
 $ Loan_Status  : chr  "N" "Y" "Y" "Y" ...


In [213]:
mod = glm(Loan_Status ~ Property_Area, data = df_1, family = binomial(link='logit'))

In [205]:
summary(mod)


Call:
glm(formula = Loan_Status ~ Property_Area, family = binomial(link = "logit"), 
    data = df_1)

Deviance Residuals: 
    Min       1Q   Median       3Q      Max  
-1.7285  -1.3693   0.7131   0.9222   0.9969  

Coefficients:
                       Estimate Std. Error z value Pr(>|z|)    
(Intercept)              0.4406     0.1713   2.572 0.010126 *  
Property_AreaSemiurban   0.7991     0.2422   3.300 0.000968 ***
Property_AreaUrban       0.1943     0.2413   0.805 0.420614    
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 609.96  on 491  degrees of freedom
Residual deviance: 597.65  on 489  degrees of freedom
AIC: 603.65

Number of Fisher Scoring iterations: 4


In [200]:
df_res = data.frame(Property_Area = c('Rural','Semiurban','Urban'))

In [195]:
nrow(df)

[1] 492

In [201]:
nrow(df_res)

[1] 3

In [214]:
predict(mod,data = df_res,type = 'response')

1         2         3         4         5         6         7         8 
0.6083916 0.6535948 0.6535948 0.6535948 0.6535948 0.6535948 0.7755102 0.6535948 
        9        10        11        12        13        14        15        16 
0.7755102 0.6535948 0.6535948 0.6083916 0.6535948 0.6535948 0.6535948 0.6083916 
       17        18        19        20        21        22        23        24 
0.6535948 0.6535948 0.7755102 0.6083916 0.7755102 0.7755102 0.6535948 0.6535948 
       25        26        27        28        29        30        31        32 
0.6535948 0.6083916 0.7755102 0.6083916 0.7755102 0.6535948 0.7755102 0.6535948 
       33        34        35        36        37        38        39        40 
0.6535948 0.7755102 0.6535948 0.6535948 0.6535948 0.7755102 0.7755102 0.7755102 
       41        42        43        44        45        46        47        48 
0.7755102 0.6535948 0.6535948 0.7755102 0.7755102 0.6083916 0.6535948 0.6535948 
       49        50        51        52        53        54        55        56 
0.6535948 0.6535948 0.6083916 0.7755102 0.7755102 0.6535948 0.6535948 0.6535948 
       57        58        59        60        61        62        63        64 
0.7755102 0.6535948 0.7755102 0.7755102 0.7755102 0.6535948 0.6535948 0.6535948 
       65        66        67        68        69        70        71        72 
0.7755102 0.7755102 0.6535948 0.6535948 0.7755102 0.7755102 0.6535948 0.7755102 
       73        74        75        76        77        78        79        80 
0.7755102 0.7755102 0.6535948 0.7755102 0.7755102 0.7755102 0.7755102 0.7755102 
       81        82        83        84        85        86        87        88 
0.7755102 0.6535948 0.7755102 0.6535948 0.6535948 0.6535948 0.7755102 0.6535948 
       89        90        91        92        93        94        95        96 
0.6083916 0.7755102 0.6083916 0.6535948 0.7755102 0.7755102 0.7755102 0.6083916 
       97        98        99       100       101       102       103       104 
0.6083916 0.6535948 0.6535948 0.7755102 0.7755102 0.7755102 0.7755102 0.6083916 
      105       106       107       108       109       110       111       112 
0.6535948 0.7755102 0.6083916 0.6083916 0.6535948 0.7755102 0.7755102 0.6535948 
      113       114       115       116       117       118       119       120 
0.7755102 0.6535948 0.6535948 0.6083916 0.7755102 0.6083916 0.6083916 0.6535948 
      121       122       123       124       125       126       127       128 
0.6083916 0.6535948 0.7755102 0.6535948 0.7755102 0.7755102 0.6535948 0.7755102 
      129       130       131       132       133       134       135       136 
0.6083916 0.6535948 0.6083916 0.6083916 0.7755102 0.7755102 0.6535948 0.6083916 
      137       138       139       140       141       142       143       144 
0.7755102 0.7755102 0.6083916 0.6083916 0.7755102 0.7755102 0.6535948 0.6535948 
      145       146       147       148       149       150       151       152 
0.7755102 0.7755102 0.7755102 0.7755102 0.6083916 0.6083916 0.6083916 0.6083916 
      153       154       155       156       157       158       159       160 
0.7755102 0.6535948 0.7755102 0.6083916 0.7755102 0.6083916 0.6535948 0.7755102 
      161       162       163       164       165       166       167       168 
0.7755102 0.6535948 0.7755102 0.7755102 0.6535948 0.6083916 0.6535948 0.7755102 
      169       170       171       172       173       174       175       176 
0.7755102 0.7755102 0.6535948 0.6083916 0.6535948 0.7755102 0.6083916 0.7755102 
      177       178       179       180       181       182       183       184 
0.6535948 0.7755102 0.7755102 0.7755102 0.6083916 0.7755102 0.6083916 0.7755102 
      185       186       187       188       189       190       191       192 
0.7755102 0.7755102 0.6083916 0.6083916 0.7755102 0.6083916 0.6535948 0.6535948 
      193       194       195       196       197       198       199       200 
0.6083916 0.7755102 0.6535948 0.6535948

In [216]:
unique(predict(mod,data = df_res,type = 'response'))

[1] 0.6083916 0.6535948 0.7755102

- Chances of getting loan against property in _Rural_ area is ~61%
- Chances of getting loan against property in _Semi urban_ area is ~77.5%
- Chances of getting loan against property in _Urban_ area is ~65%